### ISSUE TITLE
Instance segmentation feature request

### ISSUE THREAD

**Issue by mathieuboudreau** [axondeepseg/axondeepseg#662]
Per this discussion

### Discussed in https://github.com/axondeepseg/axondeepseg/discussions/661

<div type='discussions-op-text'>

<sup>Originally posted by **mcib3d** July  6, 2022</sup>
Hi and thanks for this wonderful tool,

I would eventually perform more analysis on the detected objects axons and myelins. I can  easily get a labelled image with the axons since they are separated. Sometimes two myelins touch, so my questions are : 

- how are they separated in the analysis ? It is a simple watershed ?
- can we get access to the _hidden_ separated myelins ?

Many thanks.

Thomas</div>

**Comment by hermancollin** [comment]
I will start working on this but I think what I have in mind will break retro-compatibility of the morphometrics API. More specifically, what bothers me is this:
https://github.com/axondeepseg/axondeepseg/blob/a800e307cfab899a55d7d60db2719c6f29184a95/AxonDeepSeg/morphometrics/compute_morphometrics.py#L48-L64

We already have 2 flags called `return_index_image` and `return_border_info`. In the future, if we want to add another option to this core function (i.e. `return_instance_seg` for this issue), the list will keep on getting bigger. While this will not break backwards compatibility (because they are optional parameters), it will get more and more messy everytime we try to add specific features at the morphometrics computation stage. What I would prefer is a dedicated `morphometrics_flags` data structure inside this file: [AxonDeepSeg/params.py](https://github.com/axondeepseg/axondeepseg/blob/a800e307cfab899a55d7d60db2719c6f29184a95/AxonDeepSeg/params.py)
We could keep the exact same behavior but group the flags together so they can be fed more elegantly to `get_axon_morphometrics`:
```python
get_axon_morphometrics(im_axon, path_folder, im_myelin, pixel_size, axon_shape, morpho_flags)
```
instead of:
```python
get_axon_morphometrics(im_axon, path_folder, im_myelin, pixel_size, axon_shape, return_index_image, return_border_info, return_instance_seg, return_something_else)
```

@mathieuboudreau, @Stoyan-I-A thoughts?

**Comment by hermancollin** [comment]
Ah, also, while we're at it, we should add another format to output the instance segmentation in addition to the colored map. 

The most elegant way to color these instance segmentations would be to use 4 colors, which is the mathematical minimum value of colors required for a 2D map. However, I would also like to have an intermediate representation that advanced users can manipulate easily. I'm thinking of some sort of dictionary with instance ID as key and associated pixel list as value (but more compact because we don't want to blow up the user's disk space).

**Comment by Stoyan-I-A** [comment]
> We could keep the exact same behavior but group the flags together so they can be fed more elegantly to `get_axon_morphometrics`:
> 
> `get_axon_morphometrics(im_axon, path_folder, im_myelin, pixel_size, axon_shape, morpho_flags)`

I like the idea. I also suggest having a default value for `morpho_flags`. Perhaps it could just be `get_axon_morphometrics(im_axon, path_folder, im_myelin, pixel_size, axon_shape, morpho_flags=None)`


FYI for the instance segmentation map, a long time ago, we kind of had this feature on FSLeyes, we just removed it because at the time it served no purpose, but the code is still in the `ads_plugin.py` file. The section itself is old but I used to apply a colormap called 'random' on the watershed segmentation:

```
# Apply a "random" colour mapping to the watershed mask
opts = self.displayCtx.getOpts(wantershed_mask_overlay)
opts.cmap = "random"
```
This made every value of the watershed mask have a random color. This way, adjacent axons were unlikely to have a similar color.

**Comment by mathieuboudreau** [comment]
I like the idea! Could we simply use `flags` instead of `morpho_flags`, in case we ever want to add non-morphometry related flags ? Either way works for me, just an idea.

We could also look into SCT to see if there's an even better python pattern to solve this sort of problem that they've been using. Maybe @joshuacwnewton could point us in a few directions to look at?

**Comment by joshuacwnewton** [comment]
> We could keep the exact same behavior but group the flags together so they can be fed more elegantly to `get_axon_morphometrics`:

> We could also look into SCT to see if there's an even better python pattern to solve this sort of problem that they've been using. 

Indeed, SCT has been using a similar pattern in many of its scripts. (We have objects called [`Param`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/blob/dd38d3c7d671dbc71434a4805571f69ff10a029c/spinalcordtoolbox/registration/algorithms.py#L42-L107) scattered about our codebase that contain grouped arguments.)

However, this practice was started before I joined SCT, and I'm not actually sure whether I endorse grouping arguments this way? There are some caveats I've run into when working with this pattern:

* I've found that it can sometimes make function calls more opaque and difficult to understand. Just as an example:
  
  ```python
  morpho_dict = get_axon_morphometrics(im_axon=im_in_1, 
                                       path_folder="example/folder/", 
                                       im_myelin=im_in_2, 
                                       pixel_size=5, 
                                       axon_shape="circle", 
                                       morpho_flags=example_flag_object)
  ```

   Just by looking at this call, it's unclear what parameters `morpho_flags` expects, and it's also unclear whether or not `example_flag_object` meets the necessary requirements.

  Say, then, you describe the requirements using comments inside of the docstring. Even still, you've lost the built-in argument validation given by IDEs and the Python interpreter. This means that users/devs are required to spend more time digging into documentation, then manually validating by hand that they've passed the correct parameters, which can be tedious and error-prone.

* This change can also obscure the default values of the arguments. With individual arguments, you define the default values at the function level, in the same place where you define the arguments themselves. By comparison, when you use a separate object, the defaults are defined somewhere else, so it takes more effort to understand what the default behavior is. 

Because of this, I've found that grouping arguments can introduce subtle, hard-to-notice bugs. For example, if you change one of the flag names in one place, but you forget to change the flag in another place, it's less likely to be automatically caught (since it's buried within an object). By comparison, with individual arguments, your IDE will immediately catch the mistake and likely give you a warning.

----

As an aside, I think the question here goes beyond Python-specific patterns, and touches on more general programming methodologies? Specifically, I think this ties into discussions like ["How many parameters are too many?"](https://stackoverflow.com/questions/174968/how-many-parameters-are-too-many):

> Of course, you could use a structured variable as a workaround: putting all those variables in a single struct and passing it to the routine. In fact, using structures to simplify parameter lists is one of the techniques [described by Steve McConnell in Code Complete.](https://books.google.ca/books?id=LpVCAwAAQBAJ&pg=PA322&lpg=PA322&dq=code+complete+%22Careful+programmers+avoid+bundling+data+any+more+than+is+logically+necessary.%22&source=bl&ots=GIMrmcJjPJ&sig=ACfU3U33J9M4LqUhWulrmafcbELMZEIkyw&hl=en&sa=X&ved=2ahUKEwj-4v3-1PP4AhXClWoFHWEZBuIQ6AF6BAgCEAM#v=onepage&q=code%20complete%20%22Careful%20programmers%20avoid%20bundling%20data%20any%20more%20than%20is%20logically%20necessary.%22&f=false) But as he says:
> 
> >  Careful programmers avoid bundling data any more than is logically necessary.
> 
> So if your routine has too many parameters or you use a struct to disguise a big parameter list, you're probably doing something wrong. That is, you're not keeping coupling loose.

At the same time, this quote isn't the full story -- there are all kinds of opinions in that thread. Some folks will say that having many parameters [suggests a need to refactor the function](https://stackoverflow.com/a/175041), while other folks will say that there are situations where it's [necessary to keep many parameters](https://stackoverflow.com/a/175087). 

All in all, I think it's very case-by-case and context-dependent. 

----

In ADS's case, I think having 7-8 arguments for the function is OK?

Looking at the APIs of other scientific computing libraries (like [`scikit-learn`](https://scikit-learn.org/stable/modules/classes.html)), their functions and classes often have similarly lengthy signatures with multiple boolean flags. For example, [`decomposition.PCA`](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn.decomposition.PCA):

```python
class sklearn.decomposition.PCA(n_components=None, *, copy=True, whiten=False, svd_solver='auto', tol=0.0, iterated_power='auto', n_oversamples=10, power_iteration_normalizer='auto', random_state=None)
```

Given the precedent, I don't think it would be a problem for ADS to explicitly define every `return_X` flag in the function signature?

((That's just my two cents, though. There are a lot of different opinions and factors to consider here, and I wouldn't fault you for going with the `flags` object anyway.!))

**Comment by mathieuboudreau** [comment]
Thank you so much @joshuacwnewton for your detailed opinion on this matter! @hermancollin you can have the final say on this!
